### ISSUE TITLE
Change GridsearchCV, such that when performing piecewise fitting with centering, it optimizes the "global" CV score instead of the CV score of the residuals only

### ISSUE THREAD

**Issue by kamronald** [CederGroupHub/sparse-lm#69]
**Is your feature request related to a problem? Please describe.**
Issue arises for gridsearch optimization of hyperparameters when doing piecewise fit. Assume that you have performed an 
initial fit on "initial_coefs". The second "residual" fit would be on the residual of initial fit, training the "final_coefs". When performing gridsearch on the residual fit, it would be essential to search for min(CV), such that CV includes the contributions from the first fit (ie the "global" CV). Current implementation would minimize the CV only on the residuals.

**Describe the solution you'd like**
Function that manually defines k-fold train+test sets, scans through the hyperparameters, perform residual fits on training sets, add contributions back from initial fit, and predicts on test sets. The average of test RMSEs is then the global CV RMSE for that hyperparameter. 
Can also incorporate training RMSE in this process, where we perform residual fit on entire data set, add contributions back from initial fit, and predict on all structures in training set to obtain Train RMSE.

**Describe alternatives you've considered**
There may be more efficient ways to do this, open to other ideas
![l1_residual](https://user-images.githubusercontent.com/35947132/219230021-8fa53ce9-e6d4-48ea-9cf0-880ab8804f89.png)
![l1_global](https://user-images.githubusercontent.com/35947132/219230112-b1a57746-2789-41d6-b9cc-4cf46d4fdc82.png)

**Comment by qchempku2017** [comment]
I suggest writing functionalities in the base class CVXEstimator, such that piecewise fit is supported by all estimator types in sparse-lm. 

Ronald, would you mind including some of your data in the tests so that we can test the piecewise fit?

**Comment by qchempku2017** [comment]
We can write a CompositeEstimator class to allow Piecewise fitting.

**Comment by qchempku2017** [comment]
It turns out that the issue actually arise from the choice of different train/test splitting methods. When using RepeatedKFold, a CV curve similar to the second figure is generated, even when n_repeat is set to 1:
![image](https://user-images.githubusercontent.com/28149881/220822505-04ce5043-d37d-4682-bc35-f91597775e92.png)
![image](https://user-images.githubusercontent.com/28149881/220822538-11a151b6-7cb2-4475-8418-56fcb9fe4bea.png)
However, when using KFold, which is the default in GridSearchCV, a CV curve similar to the first figure will be generated:
![image](https://user-images.githubusercontent.com/28149881/220822622-40f60c68-7cb0-4ac1-9c43-ee32baaa9bf5.png)
![image](https://user-images.githubusercontent.com/28149881/220822629-61752878-eb53-49b7-acb4-514e21a83d8b.png)

Let me check what these two splitters do in different.

**Comment by qchempku2017** [comment]
The difference turns out to be that KFold does not shuffle indices before fitting by default:
![image](https://user-images.githubusercontent.com/28149881/220823359-ce0d5b37-0396-49e9-ad46-afc53f194e01.png)
![image](https://user-images.githubusercontent.com/28149881/220823389-69de5212-6d4d-4409-a062-a85e43878bd1.png)
However, in RepeatedKFold, at each repeat the feature matrix rows are shuffled:
![image](https://user-images.githubusercontent.com/28149881/220823747-cf51cfa0-e32d-472a-984a-51df6f3c00b3.png)
![image](https://user-images.githubusercontent.com/28149881/220823622-b6b887a6-1f1a-4e25-b891-96c98a28cea5.png)
![image](https://user-images.githubusercontent.com/28149881/220823682-d294e2ef-b968-4adf-ba40-d921be824666.png)

Of course, it is definitely not correct if we do not shuffle before doing split, because structures with close indices in the feature matrix might have the same compositions, and therefore highly correlated feature vectors.

The solution is always to remember specifying argument "cv=RepeatedKFold(n_splits=5, n_repeats=xxx)' when you use GridSearchCV or LineSearchCV.

Adding a StepwiseEstimator is not needed, as the issue is not from optimizing the CV RMSE of the residuals. In principle, either optimizing the CV RMSE of the residuals or optimizing the global CV is doing the same optimization problem:

$\hat{w}^{*}_2 = \argmin_{\hat{w}_2} ||y-X_1 w_1-X_2 \hat{w}_2|| + R(\hat{w}_2)$

where $R$ is the regularization function, and $w_1$ is the intercept and the point and ewald coefficients, which are already set to be fixed in the first step.

@kamronald @lbluque, I'll close this issue with PR #71 if you agree.
